File: circular-buffer/CircularBuffer.cs
using System;
using System.Collections.Generic;

public class CircularBuffer<T>
{
    private readonly int maxCapacity;
    private readonly Queue<T> queue;

    public CircularBuffer(int capacity)
    {
        maxCapacity = capacity;
        queue = new Queue<T>(capacity);
    }

    public T Read() => queue.Count == 0
        ? throw new InvalidOperationException("No elements")
        : queue.Dequeue();

    public void Write(T value)
    {
        if (queue.Count == maxCapacity)
            throw new InvalidOperationException("Full buffer");
        
        queue.Enqueue(value);
    }

    public void Overwrite(T value)
    {
        if (queue.Count < maxCapacity)
            queue.Enqueue(value);
        else
        {
            queue.Dequeue();
            queue.Enqueue(value);
        }
    }

    public void Clear() => queue.Clear();
}

File: fsharp/circular-buffer/CircularBuffer.fs
module CircularBuffer

type CircularBuffer<'a> = 
    {
        items: 'a list;
        size: int 
    }

let mkCircularBuffer size =
    { 
        items = []
        size = size 
    }

let clear buffer =
    { buffer with items = [] }    

let write value buffer = 
    match List.length buffer.items = buffer.size with
    | true -> failwith "Cannot write to full buffer"
    | false ->{ buffer with items = buffer.items @ [value] }
        
let forceWrite value buffer =
    match List.length buffer.items = buffer.size with
    | true -> { buffer with items = List.tail buffer.items @ [value] }
    | false  -> { buffer with items = buffer.items @ [value] }

let read buffer =
    match buffer.items with    
    | x::xs -> x, { buffer with items = xs }
    | [] -> failwith "Cannot read from empty buffer" 